Upgrade to Symfony 3.0 compatibility

drewclauson opened the pull request:
For #44 

I have tested this in my environment and haven't gotten any deprecation warnings or error messages with Symfony 2.8 or 3.0.  I don't have a ton of background with upgrading bundles, but it seems to me like it'll work just by updating the dependencies.

mmarchois commented:
When does this PR merged ?

geekdevs commented:
@philippe-levan: ping, any news here?

philippe-levan commented:
Hi,
I don't have much time to work on this PR.
Can anyone check why travis is going to red and fix the problem ?
Best regards,
Philippe

drewclauson commented:
I'm not very familiar with travis at this point, but this is the only error message showing up.
```
[InvalidArgumentException]
Connection does not contain a 'path' or 'dbname' parameter and cannot be dropped.                                                                       
```

drewclauson commented:
Never mind, I wasn't reading far enough.

geekdevs commented:
@philippe-levan Why do you git ignore composer.lock? I get exactly the same errors in master even though travis shows it as passing. I believe this is because some dependencies have updated since then.

philippe-levan commented:
@geekdevs we have to ignore the composer.lock in bundles. It is the project that fixes some versions, not the included bundles.
If we have to fix some constraints on versions in a bundle we have to set these constraints in the composer.json.

Best regards,
Philippe

drewclauson commented:
The issue looks like it's how symfony 3.0 requires a string of the class name instead of the actual class when creating a form from the factory. I'll look into fixing it in the next few days unless someone else beats me to it.

geekdevs commented:
I solved `Connection does not contain a 'path' or 'dbname'` with this  PR https://github.com/TheBigBrainsCompany/TbbcMoneyBundle/pull/49
@philippe-levan Please merge it.

geekdevs commented:
@drewclauson it was merged, please rebase from master

philippe-levan commented:
@drewclauson thanks for working on this issue.

Keep in mind it would be better to stay backward compatible with ~2.1 versions if it is possible.

Otherwise I will do a new major version for this upgrade.

Thanks,
Philippe

philippe-levan commented:
@geekdevs with your PR #49 merged, tests are still failing due to an evolution of symfony/form 3.x. I is the issue that mentions @drewclauson in its comment 172937869.

Regards, Philippe

geekdevs commented:
@philippe-levan yes, I know that. But still need to rebase as otherwise it would fail on that dbpath thing.

drewclauson commented:
@philippe-levan From my research and initial testing, we are going to have to release a new major version.  I don't see a path for this to work once we hit 3.0.  How would you like me to handle this?  I'm working through the fixes today.

drewclauson commented:
All tests are now passing and I removed PHP 5.3 and 5.4 from Travis since Symfony 3.0 is >=5.5.9.  Also added in testing on PHP 7.0 for good measure.  This is definitely going to need to be a new major version - @philippe-levan anything specific I need to do in order to help make that happen?

drewclauson commented:
Update: I think it'll need to be a new major version, as I don't see an easy way to make the form changes backwards compatible.

Also, I did note that tests failed for me when I tried to run master against Symfony 2.1 or 2.2.  I think the biggest issue was the namespace of `TypeTestCase`, but I didn't look into it too closely.

geekdevs commented:
@drewclauson According to this doc http://symfony.com/doc/current/contributing/community/releases.html#schedule I think we need to support at least 2.3, 2.7, 2.8 since this are LTS versions.

Looking through your update it looks like the only major change is usage of class names for form elements instead of objects in form builder. I don't see it as being hard to support.

Also, even though v.3 is PHP>=5.5.9, we can still allow lower versions of php for v.2 branch (need to setup travis accordingly).

drewclauson commented:
@geekdevs thanks for the info, I agree with supporting all current LTS. I'm not well experienced with travis configuration. What would that configuration look like? Does travis also have the ability to test against multiple symfony versions rather than just the latest one the composer.json specifies?

geekdevs commented:
@drewclauson correct. I made some experiments with it, see example config here https://github.com/geekdevs/TbbcMoneyBundle/blob/master/.travis.yml

And that's how travis tests it https://travis-ci.org/geekdevs/TbbcMoneyBundle

drewclauson commented:
Ok, all tests passing and this PR supports Symfony >=2.3.  The readme says that it supports Symfony >= 2.1, but I'm not sure how I would make it support back to 2.1, due to the change in namespace on `TypeTestCase`.  @philippe-levan, I believe this is ready to be merged, let me know if you have any thoughts on it.

@geekdevs Thanks for your help and input.  This has been a great learning experience for me!

geekdevs commented:
@drewclauson Thanks for updates. But looks like tests are failing now. Quick check shows that this is not related to your change, but it looks quite weird and should be fixed anyway.

`PairManagerTest::testRatioProvider()` checks yahoo finance to get exchange rates. It checks it twice in the same test and values can differ between calls. I think this should be updated to be mocks.
It essentially tests that `PairManager` uses provider set with `setRatioProvider`, so provider should not  necessary be `YahooFinanceRatioProvider` (since it's tested separately), but any other class implementing that interface.

geekdevs commented:
I've created a PR for this
https://github.com/TheBigBrainsCompany/TbbcMoneyBundle/pull/50

drewclauson commented:
I was just about to look into this and saw your message.

@philippe-levan Please merge #50 and then I'll rebase and that should make this PR ready to go.

philippe-levan commented:
@drewclauson #50 merged. Thanks both (@geekdev and you) for your job on that bundle !

geekdevs commented on Tests/Form/Type/SimpleMoneyTypeTest.php:
```diff
@@ -72,12 +62,33 @@ public function testGreaterThan1000Valid()
+    protected function getExtensions()
+    {
+        //This is probably not ideal, but I'm not sure how to set up the pair manager
+        // with different decimals for different tests in Symfony 3.0
```
@drewclauson In my opinion the structure which was here previously is more explicit (i.e. you initialize money type for each test separately). The only thing you're missing is that you cannot pass params to constructor.
But you can provide them as third param of form builder (because they are defined with `setAllowedTypes`)

drewclauson replied:
@geekdevs I haven't gotten too deep into how forms work because my experience with Symfony is building a REST JSONAPI.  Can you help me out with an example of how this would work?  I agree that the structure was more explicit, but I didn't know how to pass params to the constructor without getExtensions()

geekdevs replied:
@drewclauson No worries, you made some great work!

Here's an example of how it works for  CurrencyType:

```
$form = $this->factory->create('Tbbc\MoneyBundle\Form\Type\CurrencyType', null, array(
    'reference_currency' => 'EUR',
    'currency_choices' => ['EUR','USD']
));
```

So as you can see you provide optional details as third parameter of the factory's `create` method. These parameters are accepted because of these definitions in `CurrencyType`:

```
        $resolver->setDefaults(array(
            'reference_currency' => $this->referenceCurrencyCode,
            'currency_choices' => $this->currencyCodeList
        ));
        $resolver->setAllowedTypes('reference_currency', 'string');
        $resolver->setAllowedTypes('currency_choices', 'array');
```

For `MoneyType` however it might be more difficult after symfony changes. Because currently it expects preconfigured `CurrencyType` object as an argument, which we cannot do with symfony 3. So I think we maybe need to have `reference_currency` and `currency_choices` on `MoneyType` as well and then pass it on to `CurrencyType`.

Thoughts @drewclauson @philippe-levan?

drewclauson replied:
I tried this example in CurrencyTypeTest and removed getExtensions(), but got an error message saying that I was missing the arguments for the constructor of the CurrencyType .  This was on Symfony 3.0.  Looking through the code in FormRegistry, I don't see how the options as you described would ever get to the constructor, so while I agree that we should initialize objects for each test separately, I don't see how this would work or how it would work in Symfony 3.0 at this point.  Here's a thread I had found that inspired how it's structured right now: https://github.com/symfony/symfony-docs/issues/6053

geekdevs replied:
@drewclauson I did not mean to say that options would go to constructor. Constructor should have some default settings (as nulls[?]) which would then be set as defaults via `$resolver->setDefaults()` and which could be changed when you create it with form factory with third param as shown above:

```
$form = $this->factory->create('Tbbc\MoneyBundle\Form\Type\CurrencyType', null, array(
    'reference_currency' => 'EUR',
    'currency_choices' => ['EUR','USD']
));
````

That way we would be able to initialize it both with constructor (this is what extension would do) and with form factory (useful for more explicit tests).

geekdevs commented:
@drewclauson Please also install `phpunit-bridge` (https://github.com/symfony/phpunit-bridge#usage) and run tests with it. It showed couple more deprecation errors for me that would need to be fixed.
Let me know if you'd need help fixing those (remember, it should be backwards-compatible).

drewclauson commented:
@geekdevs Unfortunately, I won't have time to engage on this again until Monday.  If you have the time to work on this before I do, feel free. :smile:

drewclauson commented:
I installed phpunit-bridge but go no deprecation warnings...

geekdevs commented:
@drewclauson Indeed, I tried it on your branch and there are no warnings. But this is strange, since e.g. `MoneyType::getName` is clearly deprecated in favor of `getBlockPrefix`. My guess this is because `getName` was completely removed from symfony3 and getBlockPrefix is auto-generated like this now:

```
    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return StringUtil::fqcnToBlockPrefix(get_class($this));
    }
```

But this is probably not what we want and might break some integrations.
So we should probably run phpunit-bridge through travis so that it checks across all versions of symfony.

geekdevs commented:
any news @drewclauson ?

drewclauson commented:
@geekdevs I've been under a lot of pressure at work to get a few deadlines done.  I may not be able to revisit this until mid-Feb. :-(

mrcmorales commented:
@geekdevs @drewclauson How is it ? Thanks :)

drewclauson commented:
@mrcmorales Unfortunately, I'm not able to return to it, but I just reviewed the last few notes and will follow up on a few questions I have.

@geekdevs How do we get phpunit-bridge into travis?  Is that in a config?

drewclauson commented:
@geekdevs Never mind, I've only drank half my :coffee: today so far... Just pushed a composer.json with phpunit-bridge included

drewclauson commented:
@geekdevs I think I've got phpunit-bridge running in Travis, but I don't see any deprecations.  The only problem is on job #170.9 and that looks probably like a Yahoo API issue.  Let me know when you can take a look.  Thanks!

geekdevs commented:
@drewclauson I've created a pull request, see https://github.com/TheBigBrainsCompany/TbbcMoneyBundle/pull/56

There are couple things why it did not show any errors (it will now):

- we always tested the same (latest) version of symfony
- somehow running `vendor/bin/phpunit` behaves differently than just `./phpunit` (only latter command actually outputs deprecation messages, versions of both are same). I've digged a little into the libraries and this looks like a bug, will need to check further though.

geekdevs commented:
@philippe-levan Please merge my PR so that we could continue testing symfony3 integration

geekdevs commented:
@drewclauson so my PR was merged, please rebase from master, set `export SYMFONY_DEPRECATIONS_HELPER=strict` at `.travis.yml` and run your tests through travis again. You will see tons of deprecation messages.

drewclauson commented:
Unfortunately, my work was re-prioritized a few weeks ago and I'm no longer working on our backend transition, which requires this library (and no one else at our company is either until we do some legacy system upgrades) but am switching gears to our legacy system... I'll leave my fork available for someone to pick up, but I won't be able to revisit this for at least 4-5 months. :frowning:

geekdevs commented:
@drewclauson normally people work on open source projects in their spare time ;). I can continue with it, no problem. Just thought you'd want to finish up yourself.
I am just not sure how it works with forks though. I assume I can fork your fork and then create PR into this repo directly?

drewclauson commented:
@geekdevs I know, I unfortunately, don't have the spare time most are able to commit to open source projects and have only been able to get more involved lately due to needing this for my job. :neutral_face:  I wish I did have the time to finish this up.

I think you can just fork my fork and create a PR back into this repo, but I haven't tried that before.  Thank you and I do apologize that I can't work on it further at this point.

geekdevs commented:
@drewclauson that's ok. You'd done a lot there already, that's much appreciated.

philippe-levan commented:
@drewclauson i do agree with @geekdevs. Thanks for your contributions ! (and I'm desperately not available neither...)

wavyx commented:
Is there anything left blocking this PR? or is it only a travis config issue?

geekdevs commented:
@wavyx it has a lot of backwards compatibility issues. Travis config fix would basically just show them in tests.  Nothing crazy complicated, but needs time to solve.

gregholland commented:
Any downside to removing backwards compatibility and tagging this as a new major version?

philippe-levan commented:
@gregholland A new major version is needed to go to SF3. But it still needs to be green in travis.
I will create a brand new 2.x branch for SF2 version and master will soon be the SF3 version.

But right now it is not compatible with SF3 and I have absolutely no time for working on this adaptation. (and I don't have any project on SF3 right now). This PR is the most advanced adaptation right now (thanks @geekdevs and @drewclauson for that).

verschoof commented:
@philippe-levan if you create a branch for SF3 we can go start testing and debugging en make PR's to solve any problem. Can you do that?

geekdevs commented:
As far as I can tell there is no big problem for making this bundle backwards compatible, so I would vote for keeping it in minor release.
The problem is more that it does not actually support v3 correctly yet. 
@drewclauson progressed there quite far, but unfortunately has no time to finish. I have intention to finish his work myself, but cannot promise on timeframes as I am quite busy as well right now.

So if any of you guys have time to finish it faster you can fork @drewclauson's branch and work from there

verschoof commented:
@geekdevs why not just merge it, and solve issue by issue. thats way more easier then fix it all right?
This would be also more easier for the community to work on.

geekdevs commented:
I am good with that approach. If @philippe-levan creates feature branch for that we could merge @drewclauson's fork and start fixing from there.

philippe-levan commented:
Hi @geekdevs, @verschoof,

I merged this PR on a public branch : "symfony3-migration"

https://github.com/TheBigBrainsCompany/TbbcMoneyBundle/tree/symfony3-migration

This feature branch will be merged on master. 

Master is now the branch for the symfony3 version of the bundle and I created a 2.x branch for the symfony 2.x version.

Do you need something else from me ?

Best regards,
Philippe

PS : I modified .travis-ci and composer.json on the symfony3-migration branch to be compatible only with sf3 and not sf2 anymore. Only the branch 2.x remains compatible with sf2.

mrcmorales commented:
hi @philippe-levan ,

Is there some version of this bundle with money v3 compatible with sf 2.7 or 2.8 ?

Thanks

philippe-levan commented:
@mrcmorales : This comment should be another issue. It is not the same problem.

I understand your are talking of the v3 of the moneyphp/money (https://github.com/moneyphp/money/tree/v3.0.0-beta) right ?

No, we won't support moneyphp/money:~3 in that bundle. There is too much BC Breaks in the library (and one reason to remain on the v2). We can't remain stable enough in our interfaces. I believe it would be a better idea to create another bundle for that moneyphp/money:~3 version.

Best regards,
Philippe

philippe-levan commented:
closed following the #issuecomment-220272885

geekdevs commented:
I've added a PR that should fix Symfony 3 integration
https://github.com/TheBigBrainsCompany/TbbcMoneyBundle/pull/65

That should support both 2.8 and 3.0
